### PR TITLE
logger abstraction and few code fixes for making diskann.dll buildabl…

### PR DESCRIPTION
**Pull request by ltan1ms**
The changes were ported from the previous [PR ](https://github.com/microsoft/DiskANN/pull/110) from my forked repo.
But send this new PR after I got permissions to push to this repo directly, and will discard the previous one.

# Changes
All changes only effects when "EXEC_ENV_OLS" is defined:
- Expose a `SetCustomLogger `function to pass in custom logging function so we can invoke it, this replaced the old undefined `DISKANNLOGGING `micro and removed dependency on some unnecessary headers, like `IANNIndex.h, ANNLogging.h` etc.
- Start with supporting log level `Info` and `Error `only as that's what `diskann::cout `or `diskan::cerr` supports.
- Added missing declarations for overloads that takes `AlighedFileReader` in `index.h`
- Added missing implementations for following function overloads that takes `AlighedFileReader` in `utils.cpp `(copied from DLSV repo): `get_bin_metadata, load_bin, copy_aligned_data_from_file, read_array, read_value`

### CODE REVIEW

File: include/index.h [hunk 1]
```diff
@@ -365,11 +374,19 @@ namespace diskann {
     DISKANN_DLLEXPORT _u64   save_data(std::string filename);
     DISKANN_DLLEXPORT _u64   save_tags(std::string filename);
     DISKANN_DLLEXPORT _u64   save_delete_list(const std::string &filename);
+#ifdef EXEC_ENV_OLS
+    DISKANN_DLLEXPORT size_t load_graph(AlignedFileReader &reade,
```
Comment (harsha-simhadri) [inline include/index.h:378]: reader instead or reade?

File: include/logger_impl.h [hunk 2]
```diff
@@ -46,7 +28,7 @@ namespace diskann {
     char*              _buf;
     int                _bufIndex;
     std::mutex         _mutex;
-    ANNIndex::LogLevel _logLevel;
+    LogLevel _logLevel;
```
Comment (harsha-simhadri) [inline include/logger_impl.h:31]: please run clang-format on files being checked in, or use VS built-in formatter that recognizes clang-format directives

File: src/utils.cpp [hunk 3]
```diff
@@ -226,4 +230,257 @@ namespace diskann {
                   << std::endl;
   }
 
+ #ifdef EXEC_ENV_OLS
+  void get_bin_metadata(AlignedFileReader& reader, size_t& npts, size_t& ndim,
```
Comment (harsha-simhadri) [inline src/utils.cpp:234]: Do you prefer to replicate most of the load code between EXEC_ENV_OLS and other environments? Or is there a way to share most of the code and avoid replication?

File: src/utils.cpp [hunk 4]
```diff
@@ -226,4 +230,257 @@ namespace diskann {
...
+                                                    float&             value,
+                                                    size_t             offset);
+  template DISKANN_DLLEXPORT void read_value<uint32_t>(
+      AlignedFileReader& reader, uint32_t& value, size_t offset);
+  template DISKANN_DLLEXPORT void read_value<uint64_t>(
+      AlignedFileReader& reader, uint64_t& value, size_t offset);
+
+#endif
+
 }  // namespace diskann
```
Comment (harsha-simhadri) [inline src/utils.cpp:486]: add new line at the end. Clang-format might fix this.

**Review by harsha-simhadri — CHANGES_REQUESTED** [review]
Li, looks good. I suggested some minor changes and formatting files inline. Would be good to get Gopal's review too

File: src/utils.cpp (hunk 3 above)
Comment (gopal-msr) [reply src/utils.cpp:234]: Apart from a few assignments, there won't be much code duplication

**Review by gopal-msr — APPROVED** [review]
I had to add the load() functions that take AlignedFileReader in the Index class because of deadline pressure. It might be worth investigating if we can remove them completely and simply use FileBlobs. But this is not necessary.

File: include/index.h [hunk 5]
```diff
@@ -14,6 +14,10 @@
 #include "tsl/robin_map.h"
 #include "tsl/sparse_map.h"
 
+#ifdef EXEC_ENV_OLS
```
[the remark this answers is not in the document]
Comment (ltan1ms) [reply include/index.h:17]: Yes, you are right, but it's only needed when EXEC_ENV_OLS is defined, so we can reduce the file size by not including it for normal case (EXEC_ENV_OLS not defined).

File: include/index.h (hunk 1 above)
Comment (ltan1ms) [reply include/index.h]: Fixed.

File: src/utils.cpp (hunk 4 above)
Comment (ltan1ms) [reply src/utils.cpp]: yes. fixed.

File: include/logger_impl.h (hunk 2 above)
Comment (ltan1ms) [reply include/logger_impl.h]: yes, i ran it on all the files. The index.cpp have some more lines of changes in formatting.

**Review by harsha-simhadri — COMMENTED** [review]
In src/utils.cpp, do you expect to maintain the loading and storing of index files for your environment seperate from the rest of the software?

**Comment by ltan1ms** [comment]
> In src/utils.cpp, do you expect to maintain the loading and storing of index files for your environment separate from the rest of the software?

@harsha-simhadri, if i understand correctly, you mean whether we will need to maintain all those overloads that take the AlignedFileReader class for our online serving environment, right? Are these stable/fundamental utility functions that are not expected to change often? If so, the effort for maintaining it should be low.

**Review by harsha-simhadri — APPROVED** [review]

**Pull request merged**